src/Raven.Server/Commercial/LicenseManager.cs

```diff
@@ -121,40 +161,138 @@ public void Initialize(StorageEnvironment environment, TransactionContextPool co
+                {
+                    cores = Math.Min(cores, _licenseStatus.MaxCores);
+
+                    var process = Process.GetCurrentProcess();
+                    SetAffinity(process, cores);
+
+                    var ratio = (int)Math.Ceiling(_licenseStatus.MaxRamInGb / (double)_licenseStatus.MaxCores);
+                    var clusterSize = GetClusterSize();
+                    var maxWorkingSet = Math.Min(_licenseStatus.MaxRamInGb / (double)clusterSize, cores * ratio);
+                    SetMaxWorkingSet(process, Math.Max(1, maxWorkingSet));
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Info("Failed to reload license limits", e);
             }
 
-            return _licenseStatus;
+            await RecalculateLicenseLimitsIfNeeded();
         }
 
-        private void ReloadLicense()
+        private int GetClusterSize()
         {
-            var license = _serverStore.LoadLicense();
-            if (license == null)
+            if (_serverStore.IsPassive())
+                return 1;
+
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
             {
-                _licenseStatus.Attributes = null;
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
-                return;
+                return _serverStore.GetClusterTopology(context).AllNodes.Count;
             }
+        }
+
+        public async Task RecalculateLicenseLimitsIfNeeded(
+            string assignedNodeTag = null, int? assignedCores = null)
+        {
+            if (_serverStore.IsLeader() == false)
+                return;
 
             try
             {
-                _licenseStatus.Attributes = LicenseValidator.Validate(license, RSAParameters);
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
+                _recalculateLicenseLimitsSemsphore.Wait();
+
+                var licenseLimits = _serverStore.LoadLicenseLimits() ?? new LicenseLimits();
+                var coresByNode = licenseLimits.CoresByNode ?? 
+                    (licenseLimits.CoresByNode = new Dictionary<string, int>());
+
+                using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    if (assignedNodeTag != null && assignedCores != null)
+                    {
+                        coresByNode[assignedNodeTag] = assignedCores.Value;
+                    }
+
+                    var allNodeTags = _serverStore.GetClusterTopology(context).AllNodes.Keys;
+                    var coresInUse = coresByNode.Sum(x => x.Value);
+                    if (allNodeTags.Count == coresByNode.Count &&
```

Comment:
What if we added a node and added a new one? That will also have the same count.